test: assert systemd-unit installation works

yajo opened the pull request:
This test demonstrates that the recipe for partitioning found in the README doesn't really work.

I push this with the hope to get upstream help on fixing the subject, make it work and fix the instructions.

Just run `nix flake check` and see that the new node won't success. These are the relevant logs:

```
automatic1 # [   14.740856] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[964]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 20: parted: command not found
automatic1 # [   14.744515] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[965]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 23: mkfs.vfat: command not found
automatic1 # [   14.747506] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[966]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 25: parted: command not found
automatic1 # [   14.976900] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[988]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 30: cryptsetup: command not found
automatic1 # [   14.981480] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[989]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 31: cryptsetup: command not found
automatic1 # [   14.985136] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[990]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 32: pvcreate: command not found
automatic1 # [   14.988748] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[991]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 33: vgcreate: command not found
automatic1 # [   14.993186] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[992]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 34: lvcreate: command not found
automatic1 # [   14.996837] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[993]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 35: mkfs.ext4: command not found
automatic1 # [   14.999860] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[994]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 37: lvcreate: command not found
automatic1 # [   15.002631] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[995]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 39: lvcreate: command not found
automatic1 # [   15.005121] 6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create[996]: /nix/store/6hhqibkfjiir8bb6mcgkh1kk59p1fcs6-create: line 40: mkfs.ext4: command not found
automatic1 # [   15.083363] systemd[1]: install-to-hd.service: Main process exited, code=exited, status=127/n/a
automatic1 # [   15.084602] systemd[1]: install-to-hd.service: Failed with result 'exit-code'.
automatic1 # [   15.085573] systemd[1]: Failed to start install-to-hd.service.
```

It seems the required executables can't be found. That's weird because I'm importing [this nixos module](https://github.com/NixOS/nixpkgs/blob/68cd1f20e109081e65f61aa14eae641bc4a98868/nixos/modules/profiles/base.nix), which should make all those scripts available.

@moduon MT-1075

yajo commented:
[Here I created the concept of a `diskoEnv`](https://github.com/nix-community/disko/pull/7/files#diff-a034057ed50b7d33bdcd694ab139870fced7ad033122e6660df5804065e477d5) to provide all needed executables in an overridable fashion.

That's probably wrong. I guess there should be a better way to do it. But in any case, what would be the preferred path to follow?

Lassulus commented on tests/test.nix:
```diff
@@ -85,13 +85,16 @@ let
-
-  nodes.machine =
-    { config, pkgs, modulesPath, ... }:
+  tsp-create = pkgs.writeScript "create" ''
+    #!${pkgs.bash}/bin/bash
```
```suggestion
  tsp-create = pkgs.writers.writeDashBin "create" ''
```

Mic92 replied:
https://github.com/nix-community/disko/pull/18 will require bash features for sure and I don't see the value of adding dash give that its not in the nixos closure and makes implementing features like this harder.

yajo replied:
The point is that systemd will refuse to execute a script without a shebang.

Probably the shebang should be added in disko directly, but that would require disko to get `pkgs` as an input. In turn I think that should be a requirement to make it really predictable, so it always uses pinned packages in the scripts. See https://github.com/nix-community/disko/pull/20#issuecomment-1225626395.

I just didn't want to touch the lib in this PR, to make sure I was testing a real problem in current disko.

Lassulus commented:
systemd services don't use packages from environment.systemPackages. you have to add them to `systemd.services.install-to-hd.path` or build a path in the shellscript with:
```
export PATH=${lib.makeBinPath [ pkgs.... ]}
```

yajo commented:
BTW after I fix this test, I think it would be very nice to provide some helper(s) to let users provide the desired environment where all binaries are included.

Or maybe this line should be changed: https://github.com/nix-community/disko/blob/9bca66ca7d2f8c9ac39d1f4a067ae45e681b87f9/lib/default.nix#L1

To something like:

```nix
{ lib, pkgs }:
```

So we can use `${pkgs.parted}/bin/parted` instead of `parted` in scripts... and the same with all other binaries that are called.

Lassulus commented:
> After latest changes, it kinda works. However I get now:
> 
> ```
> automatic1 # [   14.714282] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[975]: mkfs.vfat: unable to open /dev/disk/by-path/pci-0000:00:09.0-part1: No such file or directory
> automatic1 # [   14.722423] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[975]: mkfs.fat 4.2 (2021-01-31)
> automatic1 # [   15.153728] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1003]: Failed to open key file.
> automatic1 # [   15.174930] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1004]: Device /dev/disk/by-path/pci-0000:00:09.0-part2 is not a valid LUKS device.
> automatic1 # [   15.295794] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1005]:   No device found for /dev/mapper/crypted.
> automatic1 # [   15.323731] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1006]:   No device found for /dev/mapper/crypted.
> automatic1 # [   15.345773] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1007]:   Volume group "pool" not found
> automatic1 # [   15.347093] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1007]:   Cannot process volume group pool
> automatic1 # [   15.384782] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1008]: mke2fs 1.46.5 (30-Dec-2021)
> automatic1 # [   15.391231] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1008]: The file /dev/mapper/pool-home does not exist and no size was specified.
> automatic1 # [   15.407658] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1009]:   Volume group "pool" not found
> automatic1 # [   15.408815] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1009]:   Cannot process volume group pool
> automatic1 # [   15.428681] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1010]:   Volume group "pool" not found
> automatic1 # [   15.429825] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1010]:   Cannot process volume group pool
> automatic1 # [   15.446529] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1011]: mke2fs 1.46.5 (30-Dec-2021)
> automatic1 # [   15.449521] lwhjrmx7qgdsck8qb0d8jl606njnbr7w-wrapped[1011]: The file /dev/mapper/pool-root does not exist and no size was specified.
> automatic1 # [   15.532643] systemd[1]: install-to-hd.service: Main process exited, code=exited, status=1/FAILURE
> automatic1 # [   15.534051] systemd[1]: install-to-hd.service: Failed with result 'exit-code'.
> automatic1 # [   15.534999] systemd[1]: Failed to start install-to-hd.service.
> ```
> 
> This is a failure that's closer to what I'd expect. But I still have to fix it. laughing
> 
> Any clues?

the interesting part should be happening before. but for some reason it seems like it can't read/write into /dev/disk/by-path

yajo commented:
> the interesting part should be happening before



See the full log: [7vwqas38ip16ihl572yxrfxdrcxzahxf-vm-test-run-disko.drv.log](https://github.com/nix-community/disko/files/9415971/7vwqas38ip16ihl572yxrfxdrcxzahxf-vm-test-run-disko.drv.log)

Lassulus commented:
you could rebase and check again, stuff should be different now (and more verbose)

yajo commented:
I'm solving that rebase, but I just found out that in #22 all the preexisting partiton tests were dumped. Those tests were still valid. They were testing an LVM setup, and also I added a test for the feature I added in #6. Why were they all removed?

A good rule of thumb is that tests should only grow. Sometimes change. But never be removed. Can you restore them please so I finish the rebase properly?

Lassulus commented:
ah, yeah, I wanted to move all configurations to examples and have a test for each example. But then the commit grew to big and I just added the minimal test for the raid stuff I was building. Will probably work on tests when adding more features, sorry

yajo commented on tests/test.nix:
```diff
@@ -25,14 +48,62 @@ in makeTest' {
+    manual1.succeed("${tsp-mount}"); # verify that the command is idempotent
+
+    for node in [manual1, automatic1]:
+      node.wait_for_file("/raid")
```
If you run the tests now, you'll see `automatic1` stalls in this line. That should mean that there's something wrong when setting up a raid with a systemd unit.

Some help here please?

Lassulus replied:
I guess it gets mounted to /mnt/raid (since it's really meant for a nixos install, maybe that should be more clearer in some doc)

yajo replied:
I don't understand... According to this line, it seems like it should mount on `/raid`, right?

https://github.com/nix-community/disko/blob/6b0b20da18cdffd09f04faee7128c557bcb9f054/example/raid.nix#L52

Besides, on machine `manual1` this check succeeds.

Lassulus replied:
huh, well we prepend `/mnt` on most paths. unsure why it should succeed on manual1 https://github.com/nix-community/disko/blob/6b0b20da18cdffd09f04faee7128c557bcb9f054/default.nix#L133

yajo reviewed (commented):
Ok, no problem. Please remember to restore them when you can.

For now I continue here with the new raid config, to not having to block it.

Here's the rebase, and some other fixes. It still doesn't work, but we're getting closer.

yajo commented:
I changed it according to https://github.com/nix-community/disko/pull/20#discussion_r955804238 but it still fails.

Check the logs, it seems related to `mdmonitor.service`: [test.log](https://github.com/nix-community/disko/files/9432678/test.log)

Lassulus commented:
sadly the log is empty for me

yajo commented:
Yikes, sorry. Try now please: [test.log](https://github.com/nix-community/disko/files/9432774/test.log)

Lassulus commented:
that looks like it ran successfully? or I'm to blind to see the error :D

yajo commented:
It gets stuck on `automatic1` in `node.wait_for_file("/mnt/raid")`, so I hit Ctrl+C to abort the test after awhile. There's no log because it doesn't complete, but if you run it locally you'll see it gets stuck.

Mic92 commented:
If you "build" a nixos test you can also run it interactively in a python repl for debugging

```
nix build .#checks.x86_64-linux.your-test.driver -L
./result/bin/nixos-test-driver -I
# Now type start_all() to launch all virtual machines
# Type server.shell_interact() to get a shell
```

yajo commented:
Turns out that using RAID for this test was an overkill. It was actually working, but there was some kind of unit order conflict that didn't let it work.

With the new tests structure it became easier to use a new test with a more naive layout. It works now.

You can squash on merge if you want.

Lassulus commented:
whoops, missed this PR, will rebase this and merge after #43 is merged

Mic92 commented:
I think this now has been more or less implemented. We have installation tests for all modules: https://github.com/nix-community/disko/blob/master/tests/lib.nix
If not, let us know. Thank you for you contribution!
